abm/agent.py

```diff
@@ -0,0 +1,186 @@
+        self._time_on_patch: int = 0
+        self._time_since_last_catch: int = 0
+        self._collected_resource_last_spot: int = 0
+
+        self.resource_cluster_radius = resource_cluster_radius
+        self.success_locs = np.empty((0, 2))
+        self.failure_locs = np.empty((0, 2))
+        self.other_agent_locs = np.empty((0, 2))
+        self.is_agent: bool = True
+
+        # Output variable
+        self._collected_resource: int = 0
+        self._traveled_distance: int = 0
+
+    @property
+    def is_moving(self):
+        return self._is_moving
+
+    @property
+    def is_sampling(self):
+        return self._is_sampling
+
+    @property
+    def is_consuming(self):
+        return self._is_consuming
+
+    @property
+    def destination(self):
+        return self._destination
+
+    @property
+    def collected_resource(self):
+        return self._collected_resource
+
+    @property
+    def traveled_distance(self):
+        return self._traveled_distance
+
+    def move(self):
+        """
+        Move agent one cell closer to the destination.
+        """
+        x, y = self.pos
+        dx, dy = self._destination
+        if x < dx:
+            x += 1
+        elif x > dx:
+            x -= 1
+        if y < dy:
+            y += 1
+        elif y > dy:
+            y -= 1
+        self.model.grid.move_agent(self, (x, y))
+
+        self._traveled_distance += 1
+
+        if np.array_equal(self.pos, self._destination):
+            self._is_moving = False
+            self._is_sampling = True
+
+    def sample(self):
+        neighbors = self.model.grid.get_neighbors(
+            self.pos,
+            moore=False,
+            include_center=True,
+            radius=self.resource_cluster_radius,
+        )
+        resource_collected = False
+        self._time_on_patch += 1
+        self._is_consuming = False
+
+        for neighbor in neighbors:
+            if isinstance(neighbor, Resource) and neighbor.catch():
+                self._collected_resource += 1
+                self._collected_resource_last_spot += 1
+                resource_collected = True
+                self._is_consuming = True
+
+        if not resource_collected:
+            self._time_since_last_catch += 1
+
+        if not self.exploitation_strategy.stay_on_patch(
+            self._time_on_patch, self._time_since_last_catch
+        ):
+            self._is_sampling = False
+            self._time_since_last_catch = 0
+
+            if self._collected_resource_last_spot == 0:
+                self.add_failure_loc(self.pos)
+
+            self._collected_resource_last_spot = 0
+            self._time_on_patch = 0
+
+    def step(self):
+        if self._is_moving and not self._is_sampling:
+            self.move()
+        elif self._is_sampling and not self._is_moving:
+            self.sample()
+        else:
+            # Select a new destination
+            self._destination = self.exploration_strategy.choose_destination(
```

Comment:
I am wondering, should we not convert the destination back from ij to xy format here? This could be a subtle bug!